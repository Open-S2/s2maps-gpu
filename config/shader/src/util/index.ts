export * from './bundle';
export * from './tree';
