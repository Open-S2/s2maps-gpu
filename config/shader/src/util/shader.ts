import { Tree } from '@lezer/common';
import { ASTParser, VirtualTable, SymbolTableT, ParsedBundle, ParsedModule, ParsedModuleCache, CompressedNode } from '../types';
import { formatMurmur53, toMurmur53 } from './hash';
import { PREFIX_VIRTUAL } from '../constants';

const EMPTY_LIST = [] as any[];
const EMPTY_TABLE = {} as any;

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const makeLoadModule = <T extends SymbolTableT = any>(
  parseShader: (code: string) => Tree,
  makeASTParser: (code: string, tree: Tree, name?: string) => ASTParser<T>,
  compressAST: (code: string, tree: Tree, symbols?: string[]) => CompressedNode[],
  decompressAST: (nodes: CompressedNode[], symbols?: string[]) => Tree,
) => (
  code: string,
  name: string = 'main',
  entry?: string,
  compressed: boolean = false,
): ParsedModule => {
  if (code == null) throw new Error(`Shader code ${name} undefined`);
  if (typeof code !== 'string') throw new Error(`Shader code ${name} is not a string`);
  let tree = parseShader(code);

  const astParser = makeASTParser(code, tree, name);
  const table = astParser.getSymbolTable();
  const shake = astParser.getShakeTable(table);

  if (compressed) {
    const {symbols} = table;
    tree = decompressAST(compressAST(code, tree, symbols), symbols);
  }
  const hash = toMurmur53(code);

  return bindEntryPoint({name, code, hash, table, shake, tree}, entry);
}

// Use cache to load modules
export const makeLoadModuleWithCache = (
  loadModule: (code: string, name?: string, entry?: string, compressed?: boolean) => ParsedModule,
  defaultCache: ParsedModuleCache,
) => (
  code: string,
  name?: string,
  entry?: string,
  cache: ParsedModuleCache | null = defaultCache,
): ParsedModule => {
  if (!cache) return loadModule(code, name, entry, true);

  const hash = toMurmur53(code);
  const cached = cache.get(hash);
  if (cached) {
    return bindEntryPoint(cached, entry);
  }
  
  const module = loadModule(code, name, undefined, true);
  cache.set(hash, module);
  return bindEntryPoint(module, entry);
}

// Load a static (inert) module
export const loadStaticModule = (code: string, name: string, entry?: string) => {
  const hash = toMurmur53([code, entry]);
  return ({ name, code, hash, table: EMPTY_TABLE });
}

// Load a virtual (generated) module
export const loadVirtualModule = <T extends SymbolTableT = any>(
  virtual: VirtualTable,
  initTable: Partial<T> = EMPTY_TABLE,
  entry?: string,
  hash?: number,
  code?: string,
  key?: number,
) => {
  let symbols = initTable.symbols ?? EMPTY_LIST;

  code = code ?? `@virtual [${symbols.join(' ')}]`;
  hash = hash ?? toMurmur53(code);
  key  = key  ?? hash;

  const name = `${PREFIX_VIRTUAL}${formatMurmur53(key).slice(0, 6)}`;

  const table = {
    symbols,
    visibles: symbols,
    ...initTable,
  };
  return { name, code, hash, table, entry, virtual, key };
}

// Set entry point of a module, returns new bundle/module.
// Is the same instance as the original (key = old key/hash), so it merges with copies of itself.
// But is structurally different (hash = new key), so differences in links are reflected in the shader hash.
export const bindEntryPoint = <T extends ParsedBundle | ParsedModule>(bundle: T, entry?: string): T => {
  let {key, hash, module, table} = bundle as any;

  table = table ?? module?.table;
  hash = hash ?? module?.hash;

  if (entry == null && table.symbols?.includes('main')) entry = 'main';
  if (entry == null) return bundle;

  const structural = toMurmur53([hash, entry]);
  return {...bundle, entry, hash: structural, key: key ?? hash};
};
