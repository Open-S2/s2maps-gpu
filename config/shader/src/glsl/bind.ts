import { ShaderModule, LambdaSource, StorageSource, TextureSource, DataBinding } from './types';

import { defineConstants } from './shader';
import { makeBindingAccessors, makeUniformBlock } from './gen';
import { makeResolveBindings, namespaceBinding } from '../util/bind';
import { VIRTUAL_BINDGROUP, VOLATILE_BINDGROUP } from './constants';

export { bindBundle, bindModule } from '../util/bind';

const NO_SYMBOLS = [] as any[];

const getVirtualBindGroup = () => VIRTUAL_BINDGROUP;

export const bindingToModule = (
  binding: DataBinding,
): ShaderModule => {
  const {uniform: {name}} = binding;
  const links = makeBindingAccessors([binding], VIRTUAL_BINDGROUP, VOLATILE_BINDGROUP);
  const module = links[name];
  return {...module, entry: name};
}

export const bindingsToLinks = (
  bindings: DataBinding[],
): Record<string, ShaderModule> => {
  return makeBindingAccessors(bindings, VIRTUAL_BINDGROUP, VOLATILE_BINDGROUP);
}

export const sourceToModule = <T>(
  source: ShaderModule | LambdaSource<T> | StorageSource | TextureSource | any,
): ShaderModule | null => {
  if (source == null) return null;

  const s = source as any;
  if (s.shader) return s.shader as ShaderModule;
  else if (s.table || s.libs) return source as ShaderModule;
  return null;
}

export const resolveBindings = makeResolveBindings(makeUniformBlock, getVirtualBindGroup);
