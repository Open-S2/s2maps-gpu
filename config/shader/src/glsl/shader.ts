import { Tree } from '@lezer/common';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { makeTranspile } from '../util/transpile';

import { makeASTParser, compressAST, decompressAST } from './ast';
import { toTypeString, toTypeArgs } from './type';
import { parser } from './grammar/glsl';
import LRU from 'lru-cache';
import zip from 'lodash/zip';

export { loadStaticModule, loadVirtualModule, bindEntryPoint } from '../util/shader';

/** LRU cache for parsed shader code */
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<number, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

/** Parse GLSL code into lezer tree */
export const parseShader = (code: string): Tree => parser.parse(code);

/** Parse a code module into its in-memory representation (AST + symbol table) */
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST, decompressAST);

/** Use cache to load modules */
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

/** Make GLSL definitions */
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

/** Make GLSL enable definitions */
export const defineEnables = (enabled: string[]) => enabled.map(e => `#extension ${e}â€‹ : enable`).join('\n');

/** Convert a bundle with a defined entry point to a definition for that attribute or type. */
export const bundleToAttribute = makeBundleToAttribute(toTypeString, toTypeArgs);

/** Convert a bundle to a definition for all its attributes. */
export const bundleToAttributes = makeBundleToAttributes(toTypeString, toTypeArgs);

// Simple whitespace removal
const minifyCode = (code: string) => {
  code = code.replace(/ +/g, ' ');
  code = code.replace(/\n+/g, '\n');
  return code;
};

/** ES/CommonJS Transpiler */
export const transpileGLSL = makeTranspile('glsl', 'glsl', loadModule, compressAST, minifyCode);

/** Templated literal syntax:

```tsx
glsl`...`
``` */
export const glsl = (literals: TemplateStringsArray, ...tokens: string[]) => {
  const code = zip(literals, tokens).flat();
  return loadModuleWithCache(code.join(''));
};

/** Format `number` as GLSL `float` */
export const float = (x: number) => {
  const s = x.toString();
  return (!s.match(/\./)) ? s + '.0' : 0;  
};
/** Format `number` as GLSL `uint` */
export const uint = (x: number) => Math.round(x).toString();
/** Format `number` as GLSL `int` */
export const int = (x: number) => Math.round(x).toString();
