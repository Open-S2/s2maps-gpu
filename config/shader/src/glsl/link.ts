import { Tree } from '@lezer/common';

import { defineConstants, defineEnables, loadModule, loadModuleWithCache, DEFAULT_CACHE } from './shader';
import { rewriteUsingAST } from './ast';
import { makeLinker, makeLinkCode, makeLinkBundle, makeLinkModule } from '../util/link';

import { GLSL_VERSION } from './constants';

// Override GLSL version/prefix
let PREAMBLE = `#version ${GLSL_VERSION}`;
export const setPreamble = (s: string): string => PREAMBLE = s;
export const getPreamble = (): string => PREAMBLE;
const getPreambles = (): string[] => [getPreamble()];

// No preprocessor renames
const getRenames = () => new Map<string, string>();

export const linker     = makeLinker(getPreambles, getRenames, defineConstants, defineEnables, rewriteUsingAST);
export const linkBundle = makeLinkBundle(linker);
export const linkModule = makeLinkModule(linker);
export const linkCode   = makeLinkCode(linker, loadModuleWithCache, DEFAULT_CACHE);
