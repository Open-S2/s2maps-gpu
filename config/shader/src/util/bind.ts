import { ParsedBundle, ParsedModule, ShaderModule, ShaderDefine, DataBinding } from '../types';

import { parseLinkAliases } from '../util/link';
import { toHash, formatMurmur53, toMurmur53, scrambleBits53, mixBits53 } from '../util/hash';
import { toBundle, toModule, getBundleHash, getBundleKey } from '../util/bundle';
import { loadStaticModule } from '../util/shader';
import { PREFIX_CLOSURE, PREFIX_VIRTUAL, VIRTUAL_BINDINGS } from '../constants';

import { timed } from './timed';

const NO_SYMBOLS = [] as any[];
const DEBUG = false;

export type BindBundle = (
  bundle: ShaderModule,
  linkDefs?: Record<string, ShaderModule | null>,
  defines?: Record<string, ShaderDefine> | null,
  key?: string | number,
) => string;

export type BindModule = (
  main: ParsedModule,
  libs?: Record<string, ShaderModule>,
  linkDefs?: Record<string, ShaderModule | null>,
  defines?: Record<string, ShaderDefine> | null,
  virtual?: ParsedModule[],
  key?: string | number,
) => ParsedBundle;

export type DefineConstants = (
  defines: Record<string, ShaderDefine>
) => string;

export type MakeUniformBlock = (
  constants: DataBinding[],
  set?: number | string,
  binding?: number | string,
) => string;

export const bindBundle = (
  subject: ShaderModule,
  links: Record<string, ShaderModule | null> | null = null,
  defines: Record<string, ShaderDefine> | null = null,
): ParsedBundle => {
  const bundle = toBundle(subject);
  if (!links && !defines) return bundle;

  const hash = getBundleHash(bundle);
  const key = getBundleKey(bundle);

  // External hash
  let external: number = 0;
  for (const k in links) if (links[k]) external = mixBits53(external, getBundleHash(links[k]!));

  const defs = defines ? toMurmur53(defines) : 0;
  const code = `@closure`;
  const rehash = scrambleBits53(mixBits53(hash, mixBits53(external, defs)));

  // External key
  external = 0;
  for (const k in links) if (links[k]) external = mixBits53(external, getBundleKey(links[k]!));
  const rekey = scrambleBits53(mixBits53(mixBits53(hash, key), mixBits53(external, defs)));

  const relinks = bundle.links ? {
    ...bundle.links,
  } : {} ?? undefined;

  const redefines = defines && bundle.defines ? {
    ...bundle.defines,
    ...defines,
  } : defines ?? bundle.defines ?? undefined;

  let revirtuals = bundle.virtuals ? bundle.virtuals.slice() : [];

  const {module: {table: {linkable}}} = bundle;
  if (links && linkable) for (const k in links) if (links[k]) {
    const chunk = links[k] as any;

    // Ensure link exists in module
    let check = k.indexOf(':') > 0 ? k.split(':')[0] : k;
    if (!linkable[check]) continue;

    // Copy bundle's sub-virtuals
    if (chunk.virtuals) revirtuals.push(...chunk.virtuals);

    // Add virtual module to list
    if (chunk.virtual) revirtuals.push(chunk);
    if (chunk.module?.virtual) revirtuals.push(chunk.module);

    relinks[k] = links[k]!;
  } else if (relinks[k]) {
    const chunk = relinks[k] as any;

    if (chunk.virtuals) revirtuals = revirtuals.filter(f => !chunk.virtuals.includes(f));

    if (chunk.virtual) revirtuals = revirtuals.filter(f => f !== chunk);
    if (chunk.module?.virtual) revirtuals = revirtuals.filter(f => f !== chunk.module);

    delete relinks[k];
  }

  return {
    ...bundle,
    links: relinks,
    defines: redefines,
    hash: rehash,
    key: rekey,
    virtuals: revirtuals,
    attributes: undefined,
  } as any;
};

export const bindModule = bindBundle;

export const makeResolveBindings = (
  makeUniformBlock: MakeUniformBlock,
  getVirtualBindGroup: (defines?: Record<string, ShaderDefine>) => string | number,
) => timed('resolveBindings', (
  modules: (ParsedBundle | null)[],
  defines?: Record<string, ShaderDefine>,
  lazy?: boolean,
): {
  modules: (ParsedBundle | null)[],
  uniforms: DataBinding[],
  bindings: DataBinding[],
  volatiles: DataBinding[],
  visibilities: Map<DataBinding, GPUShaderStageFlags>,
} => {
  const allUniforms  = [] as DataBinding[];
  const allBindings  = [] as DataBinding[];
  const allVolatiles = [] as DataBinding[];

  const allVisibilities = new Map<DataBinding, GPUShaderStageFlags>();

  const seen = new Set<number>();
  DEBUG && console.log('------------')

  const addBinding = (b: DataBinding, slots: number, visibility: GPUShaderStageFlags) => {
    const s = (b.storage ?? b.texture) as any;
    if (s && s.volatile) {
      allVolatiles.push(b);
      volatileBase += slots;
    }
    else {
      allBindings.push(b);
      bindingBase += slots;
    }
    addVisibility(b, visibility);
  }

  const addVisibility = (b: DataBinding, visibility: GPUShaderStageFlags) => {
    allVisibilities.set(b, (allVisibilities.get(b) || 0) | visibility);
  }
  
  // Gather all namespaced uniforms and bindings from all virtual modules.
  // Assign base offset to each virtual module in-place.
  let bindingBase = 0;
  let volatileBase = 0;
  let index = 0;
  let stage = 0;
  for (const m of modules) if (m) {
    const {virtuals} = m;

    const visibles = new Set<number>();
    const visibility = modules.length === 2
      ? (stage ? GPUShaderStage.FRAGMENT : GPUShaderStage.VERTEX)
      : GPUShaderStage.COMPUTE;

    if (virtuals) for (const m of virtuals) {
      const key = getBundleKey(m);

      if (seen.has(key)) {
        if (visibles.has(key)) continue;
        visibles.add(key);

        if (m.virtual) {
          const {storages, textures} = m.virtual;
          if (storages) for (const b of storages) addVisibility(b, visibility);
          if (textures) for (const b of textures) addVisibility(b, visibility);
        }        

        continue;
      }
      seen.add(key);
      visibles.add(key);

      DEBUG && console.log('virtual', m.code, m.hash, m.key);

      if (m.virtual) {
        const {uniforms, storages, textures} = m.virtual;

        // Mutate virtual modules as they are ephemeral
        const namespace = `${PREFIX_VIRTUAL}${++index}_`;
        if (!lazy) {
          m.virtual.namespace = namespace;
          m.virtual.bindingBase = bindingBase;
          m.virtual.volatileBase = volatileBase;
        }

        if (uniforms) for (const u of uniforms) allUniforms.push(namespaceBinding(namespace, u));
        if (storages) for (const b of storages) addBinding(b, 1, visibility);
        if (textures) for (const b of textures) addBinding(b, 1 + +!!(b.texture!.sampler && (b.uniform!.args !== null)), visibility);
      }
    };
    stage++;
  }

  let out = modules;

  // Create combined uniform block as top-level import
  if (!lazy) {
    const virtualBindGroup = getVirtualBindGroup(defines);
    const code = makeUniformBlock(allUniforms, virtualBindGroup, bindingBase);
    const lib = loadStaticModule(code, VIRTUAL_BINDINGS);

    const imported = {at: -1, symbols: NO_SYMBOLS, name: VIRTUAL_BINDINGS, imports: NO_SYMBOLS};

    // Append to modules
    out = modules.map((m: ParsedBundle | null) => {
      if (!m) return null;

      const bundle = toBundle(m);

      const {module, libs} = bundle;
      const {table} = module;
      const {modules} = table;

      const retable = {
        ...table,
        modules: modules ? [...modules, imported] : [imported],
      };

      return {
        ...bundle,
        module: {
          ...module,
          table: retable,
        },
        libs: {
          ...libs,
          [VIRTUAL_BINDINGS]: lib,
        },
      };
    });
  }
  
  DEBUG && console.log('visibility', allVisibilities);

  return {
    modules: out,
    uniforms: allUniforms,
    bindings: allBindings,
    volatiles: allVolatiles,
    visibilities: allVisibilities,
  };
});

export const namespaceBinding = (namespace: string, binding: DataBinding) => {
  const {uniform} = binding;
  const {name} = uniform;
  const imp = namespace + name;
  return {...binding, uniform: {...uniform, name: imp}};
};

const VIRTUAL = 'VIRTUAL';

export const getBindingArgument = (binding?: ShaderDefine): string | number => {
  if (typeof binding === 'string') {
    const a = binding.indexOf('(');
    const b = binding.indexOf(')');
    return binding.slice(a + 1, b);
  }
  if (typeof binding === 'number') return binding;
  return VIRTUAL;
}
