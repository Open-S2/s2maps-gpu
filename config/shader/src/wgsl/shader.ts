import { Tree } from '@lezer/common';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { makeTranspile } from '../util/transpile';

import { makeASTParser, compressAST, decompressAST } from './ast';
import { toTypeString, toTypeArgs } from './type';
import { removeComments, removeWhiteSpace, renameLocals } from './minify';
import { parser } from './grammar/wgsl';

import LRU from 'lru-cache';
import zip from 'lodash/zip';

export { loadStaticModule, loadVirtualModule, bindEntryPoint } from '../util/shader';

/** LRU cache for parsed shader code */
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<number, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

/** Parse WGSL code into lezer tree */
export const parseShader = (code: string): Tree => parser.parse(code);

/** Parse a code module into its in-memory representation (AST + symbol table) */
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST, decompressAST);

/** Use cache to load modules */
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

/** Make WGSL constant definitions */
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (k[0] !== '@' && defs[k] != null) out.push(`const ${k} = ${defs[k]};`);
  return out.join("\n");
}

/** Make WGSL constant definitions */
export const defineEnables = (enabled: string[]) => enabled.length ? `enable ${enabled.join(', ')};` : '';

/** Convert a bundle with a defined entry point to a definition for that attribute or type. */
export const bundleToAttribute = makeBundleToAttribute(toTypeString, toTypeArgs);

/** Convert a bundle to a definition for all its attributes. */
export const bundleToAttributes = makeBundleToAttributes(toTypeString, toTypeArgs);

// Simple whitespace / comment removal
const minifyCode = (code: string) => {
  code = removeComments(code);
  code = renameLocals(code);
  code = removeWhiteSpace(code);
  return code;
};

/** ES/CommonJS Transpiler */
export const transpileWGSL = makeTranspile('wgsl', 'wgsl', loadModule, compressAST, minifyCode);

/** Templated literal syntax:

```tsx
wgsl`...`
``` */
export const wgsl = (literals: TemplateStringsArray, ...tokens: string[]) => {
  const code = zip(literals, tokens).flat();
  return loadModuleWithCache(code.join(''));
};

/** Format `number` as WGSL `f32` */
export const f32 = (x: number) => {
  const s = x.toString();
  return (!s.match(/\./)) ? s + '.0' : s;  
};
/** Format `number` as WGSL `u32` */
export const u32 = (x: number) => Math.round(x).toString() + 'u';
/** Format `number` as WGSL `i32` */
export const i32 = (x: number) => Math.round(x).toString();
