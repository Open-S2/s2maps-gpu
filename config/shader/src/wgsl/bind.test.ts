import { bindBundle, bindingsToLinks, resolveBindings } from './bind';
import { loadModule } from './shader';
import { linkBundle } from './link';
import { formatAST } from '../util/tree'; 
import { addASTSerializer } from '../test/snapshot';

addASTSerializer(expect);

describe("bind", () => {
  
  it("bind produces a bound bundle", () => {
    
    const codeMain = `
    @link fn getColor() -> vec4<f32> {};
    fn main() {
      var v: vec4<f32>;
      v = getColor();
      return v;
    }
    `

    const codeSub = `
    @link fn getSubColor() -> vec4<f32> {};

    fn colorUsed() -> vec4<f32> { return vec4<f32>(0.0, 0.1, 0.2, 0.0); }
    fn colorNotUsed() -> vec4<f32> { return vec4<f32>(0.0, 0.1, 0.2, 1.0); }

    @export fn getColor() -> vec4<f32> {
      return getSubColor() + colorUsed();
    }
    `

    const codeColor = `
    @export fn getColor() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    `

    const module = loadModule(codeMain, 'main');
    const sub = loadModule(codeSub, 'sub');
    const getColor = loadModule(codeColor, 'getColor');

    const links = {"getSubColor:getColor": getColor};
    const defines = {'TEST': true};

    const bound = bindBundle(sub, links, defines);

    expect(bound.hash).not.toEqual(sub.table.hash);
    expect(bound.hash).toBeTruthy();
    expect(bound.links).toEqual(links);
    expect(bound.defines).toEqual(defines);

    expect(bound.hash).toMatchSnapshot();
  });
  
  it("binds an external", () => {
    
    const codeMain = `
    @link fn getColor() -> vec4<f32> {};
    fn main() {
      var v: vec4<f32>;
      v = getColor();
      return v;
    }
    `

    const codeSub = `
    @link fn getSubColor() -> vec4<f32> {};

    fn colorUsed() -> vec4<f32> { return vec4<f32>(0.0, 0.1, 0.2, 0.0); }
    fn colorNotUsed() -> vec4<f32> { return vec4<f32>(0.0, 0.1, 0.2, 1.0); }

    @export fn getColor() -> vec4<f32> {
      return getSubColor() + colorUsed();
    }
    `

    const codeColor = `
    @export fn getColor() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    `

    const module = loadModule(codeMain, 'main');
    const sub = loadModule(codeSub, 'sub', 'getColor');
    const getColor = loadModule(codeColor, 'getColor');

    const links = {"getSubColor:getColor": getColor};
    const defines = {'TEST': true};

    {
      const linked = linkBundle(sub, links, defines);
      expect(linked).toMatchSnapshot();
    }

    {
      const bound = bindBundle(sub, links, defines);
      const linked = linkBundle(bound);
      expect(linked).toMatchSnapshot();
    }

    {
      const bound = bindBundle(sub, links, defines);
      const linked = linkBundle(module, {getColor: bound});
      expect(linked).toMatchSnapshot();
    }

  });

  it('makes deterministic links for data bindings', () => {
    const dataBindings = [
      {
        uniform: { format: 'vec4<f32>', name: 'getColor', value: [0, 0.5, 1, 1], args: ['i32'] },
        constant: [1, 0.5, 1, 1],
      },
      {
        uniform: { format: 'vec2<f32>', name: 'getSize', value: [1, 1], args: ['i32'] },
        storage: {
          buffer: {} as any,
          format: 'vec2<f32>',
          length: 10,
          size: [10],
          version: 1,
        },
      },
    ];

    const toSnapshot = (link: any) => {
      const { name, code, table, virtual: { uniforms, storages, textures, base }} = link;
      return { name, code, table, uniforms, storages, textures, base };
    }

    const links1 = bindingsToLinks(dataBindings);
    expect(toSnapshot(links1.getColor)).toMatchSnapshot();

    const links2 = bindingsToLinks(dataBindings);
    expect(toSnapshot(links2.getColor)).toEqual(toSnapshot(links1.getColor));
  });

  it('links data bindings', () => {
    const dataBindings = [
      {
        uniform: { format: 'vec4<f32>', name: 'getColor', value: [0, 0.5, 1, 1], args: ['i32'] },
        constant: [1, 0.5, 1, 1],
      },
      {
        uniform: { format: 'vec2<f32>', name: 'getSize', value: [1, 1], args: ['i32'] },
        storage: {
          buffer: {} as any,
          format: 'vec2<f32>',
          length: 10,
          size: [10],
          version: 1,
        },
      },
    ];

    const code = `
    @link fn getColor(i: i32) -> vec4<f32> {};
    @link fn getSize(i: i32) -> vec2<f32> {};

    fn main() {
      getColor(0);
      getSize(0);
    }
    `;

    const mod = loadModule(code, 'code');

    const toSnapshot = (link: any) => {
      const { name, code, table, virtual: { uniforms, storages, textures, base }} = link;
      return { name, code, table, uniforms, storages, textures, base };
    }

    const links = bindingsToLinks(dataBindings);
    const bound = bindBundle(mod, links);

    const fail = () => linkBundle(bound);
    expect(fail).toThrow();

    const defines = {'@group(VIRTUAL)': '@group(0)'};
    const {modules: [resolved], uniforms, bindings} = resolveBindings([bound], defines);
    const result = linkBundle(resolved, {}, defines);
    expect(result).toMatchSnapshot();
  });

});
