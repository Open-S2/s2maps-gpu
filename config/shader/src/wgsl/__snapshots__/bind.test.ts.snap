// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bind bind produces a bound bundle 1`] = `1068531536290827`;

exports[`bind binds an external 1`] = `
"const TEST = true;

    fn _01_getColor() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    

    fn colorUsed() -> vec4<f32> { return vec4<f32>(0.0, 0.1, 0.2, 0.0); }
fn getColor() -> vec4<f32> {
      return _01_getColor() + colorUsed();
    }
    "
`;

exports[`bind binds an external 2`] = `
"const TEST = true;

    fn _01_getColor() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    

    fn colorUsed() -> vec4<f32> { return vec4<f32>(0.0, 0.1, 0.2, 0.0); }
fn getColor() -> vec4<f32> {
      return _01_getColor() + colorUsed();
    }
    "
`;

exports[`bind binds an external 3`] = `
"const TEST = true;

    fn _01_getColor() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    

    fn _02_colorUsed() -> vec4<f32> { return vec4<f32>(0.0, 0.1, 0.2, 0.0); }
fn _02_getColor() -> vec4<f32> {
      return _01_getColor() + _02_colorUsed();
    }
    

    fn main() {
      var v: vec4<f32>;
      v = _02_getColor();
      return v;
    }
    "
`;

exports[`bind links data bindings 1`] = `
"
struct _VT_Type {
  _VT_1_getColor: vec4<f32>,
};
@group(0) @binding(1) var<uniform> _VT_Uniform: _VT_Type;


fn _VT_1_getColor(a: i32) -> vec4<f32> {
  return _VT_Uniform._VT_1_getColor;
}


@group(0) @binding(0) var<storage> _VT_1_getSizeStorage: array<vec2<f32>>;

fn _VT_1_getSize(a: i32) -> vec2<f32> {
  return _VT_1_getSizeStorage[a];
}


    fn main() {
      _VT_1_getColor(0);
      _VT_1_getSize(0);
    }
    "
`;

exports[`bind makes deterministic links for data bindings 1`] = `
Object {
  "base": undefined,
  "code": "@access [getColor getSize] [bxd6ge0qaz]",
  "name": "_VT_171w40",
  "storages": Array [
    Object {
      "storage": Object {
        "buffer": Object {},
        "format": "vec2<f32>",
        "length": 10,
        "size": Array [
          10,
        ],
        "version": 1,
      },
      "uniform": Object {
        "args": Array [
          "i32",
        ],
        "format": "vec2<f32>",
        "name": "getSize",
        "value": Array [
          1,
          1,
        ],
      },
    },
  ],
  "table": Object {
    "exports": Array [
      Object {
        "flags": 1,
        "func": Object {
          "name": "getColor",
          "parameters": Array [
            "i32",
          ],
          "type": Object {
            "name": "vec4<f32>",
          },
        },
      },
      Object {
        "flags": 1,
        "func": Object {
          "name": "getSize",
          "parameters": Array [
            "i32",
          ],
          "type": Object {
            "name": "vec2<f32>",
          },
        },
      },
    ],
    "modules": undefined,
    "symbols": Array [
      "getColor",
      "getSize",
    ],
    "visibles": Array [
      "getColor",
      "getSize",
    ],
  },
  "textures": Array [],
  "uniforms": Array [
    Object {
      "constant": Array [
        1,
        0.5,
        1,
        1,
      ],
      "uniform": Object {
        "args": Array [
          "i32",
        ],
        "format": "vec4<f32>",
        "name": "getColor",
        "value": Array [
          0,
          0.5,
          1,
          1,
        ],
      },
    },
  ],
}
`;
